fix(offlinemsg): Fix offline message dispatching on history load, and move old stuck faux_offline_pending messages to a new table

anthonybilinski opened the pull request:
* Fix callback hookup order in ChatHistory
* Add correct call to SessionChatLog to insert an unfinished message
* Fix incorrect logic in parsing History database response

This is a rebase of https://github.com/qTox/qTox/pull/5735. Review will continue here now that it's rebased.

Note that it's rebased on top of yet unmerged https://github.com/qTox/qTox/pull/5772 - without which all pending messages from all friends will be sent to every friend on qTox start!

- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5773)
<!-- Reviewable:end -->

anthonybilinski committed:
fix(db): prepare and execute queued statements one at a time

* Allows for queued preparations to depend on the execution of previous statements

sphaerophoria committed:
fix(offlinemsg): fix offline message dispatching on history load

* Fix callback hookup order in ChatHistory
* Add correct call to SessionChatLog to insert an unfinished message
* Fix incorrect logic in parsing History database response

anthonybilinski commented:
Addressing https://github.com/qTox/qTox/pull/5735#pullrequestreview-256553969:
Originally from @Diadlo 
>I don't understand, why you don't want to mock History? Because it can be the root of some bugs? If so, it doesn't matter for ChatHistory test. But how to test History it's interesting question

Chatted with @sphaerophoria and he basically agrees with what you've said Diadlo. History is a more likely cause of bugs, but that doesn't mean it can't be mocked to test ChatHistory separately. Personally, I'd rather not spend time on unit testing ChatHistory right now. Is it acceptable to include the fix without?

>Looks like the first line of if and else are the ~same
auto chatLogMessage = ChatLogMessage{isComplete, processedMessage};

Agreed, fixed.

From https://github.com/qTox/qTox/pull/5735#pullrequestreview-256554895
> const reference for this function and the previous

done for sender, senderName, message, and file.

> Why now it can't be null?

I think this change was due to the fact the `message` can never be null, due to the table's definition: `message BLOB NOT NULL`.

https://github.com/qTox/qTox/pull/5735#issuecomment-517859591
From @sudden6 
> @sphaerophoria can you rebase?

Done.

Diadlo reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773#-:-LlqIO7w7gBpFOVF5MI1:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

sudden6 commented:
@anthonybilinski was the source of that "sending unsent messages to all friends" bug the database logic? Can we protect against this in the future?

anthonybilinski commented:
@sudden6 it was a combination of:
1. Bad SQL select since the initial refactor
2. Offline message sending in load fixed, causing 1. to actually have an effect.
3. me stupidly running `gdb qtox` in my build directory and thinking it would run /use/local/qtox instead of ./qtox.

As soon as the SQL select is complete, we have no way to know that the messages aren't for the friend we're sending them to, so I think the only option would be to have tests for history with some kind of mocked rawdatabase (or just a real rawdatabase like I did for the DB schema version upgrade test), and then to test each History SQL interaction on a known DB state and make sure we get back the expected date.

I think this would be worth doing in general, I some changes in progress WRT unit testing db schema logic already that could be extended.

sudden6 commented:
Ok, thanks for the explanation.

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773#-:-LlrvsPF2yIjg0k8I2Yo:b-jbsf8z)* status: :shipit: complete! 2 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

anthonybilinski commented:
Fixing this causes some unexpected behaviour due to old messages stuck as pending suddenly being sent. https://github.com/qTox/qTox/issues/5776 created to address this. I'll hold off merging this PR until a choice is made and implemented (if needed).

sphaerophoria commented:
Just linking this back to PR #5734 so it's obvious it's being worked on

anthonybilinski committed:
refactor(test): allow generic schema creation and verification in for db schema

anthonybilinski committed:
fix(history): move stuck pending message into broken_messages table

Fix #5776

Due to a long standing bug, faux offline message have been able to become stuck
going back years. Because of recent fixes to history loading, faux offline
messages will correctly all be sent on connection, but this causes an issue of
long stuck messages suddenly being delivered to a friend, out of context,
creating a confusing interaction. To work around this, this upgrade moves any
faux offline messages in a chat that are older than the last successfully
delivered message, indicating they were stuck, to a new table,
`broken_messages`, preventing them from ever being sent in the future.

anthonybilinski committed:
refactor(history): apply db schema upgrade one version at a time

This allows upgrade steps to query the db at the last version and run C++ code
on the results, then do a single transaction to make the upgrade, instead of
all actions of each upgrade step being required to be part of the overall
upgrade transaction.

anthonybilinski commented:
I've added my WIP changes for https://github.com/qTox/qTox/issues/5776, so that the fixing of sending can be merged at the same time as the db "upgrade" to avoid sending old history.

anthonybilinski committed:
fix(history): handle errors during db upgrade

anthonybilinski committed:
fix(db): show full sqlite error message

anthonybilinski commented:
This is feature complete now. I've done some preliminary testing and it looks good, but I'm going to do some more testing with crafted databases. **Careful of testing this on your database!** After some git history cleanup and more testing by me, this should be good to go for review.

anthonybilinski committed:
fix(history): select broken messages from History, track in ChatLogMessage

anthonybilinski committed:
fix(history): display broken messages UI with error icon

Allows users to still see their broken messages in the chatlog, but in a way
that's clear that they haven't been delivered and won't be sent.

anthonybilinski committed:
refactor(history): rename pending message variables

"Sent" to "Delivered", since that's what we're tracking also rename "isOffline"
to "isPending", which is boolean the same, but "isOffline" was opposite of its
correct value before, so boolean reverse the assignment.

anthonybilinski committed:
refactor(history): represent message state as enum class

ensures that only one state is set at a time, avoids managing multiple bools

anthonybilinski commented:
This is ready to be reviewed now. I made an effort to split it into logical chunks, so I recommend reviewing per commit.

* The RawDatabase change was cherry-picked from master, originally merged as https://github.com/qTox/qTox/pull/5842, because the corrected sql transaction behaviour was needed for the db upgrade to prepare statements in the transaction based off the creation of the new table earlier in that same transaction.
* the "fix(offlinemsg): fix offline message dispatching on history load" change is what was already reviewed for this PR, unchanged.
* the 8 commits after that are the new changes, to fix #5776

sudden6 reviewed (changes requested):
> Reviewed 5 of 5 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773)* status: :shipit: complete! 2 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 187 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquL5QaEZW_qGOEGvpU:-LquL5QaEZW_qGOEGvpV:b-xnkkk6) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/src/persistence/history.cpp#L187)):*
> ```cpp
>         //   possibility 2) it is a old existing database, before version 1 and before we saved schema version,
>         //       and needs to be updated.
>         bool bSuccess;
> ```

please initialize the variable
___
*[src/persistence/history.cpp, line 670 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquPv0-8bkJSxXL-PE4:-LquPv00AKFXCRot4Ci8:bc1u22l) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/src/persistence/history.cpp#L670)):*
> ```cpp
> 
>         assert(!(isPending && isBroken));
>         MessageState messageState;
> ```

There are multiple copies of the conversion, please extract into a function or make a constructor for MessageState (if C++ supports that)
___
*[test/persistence/dbschema_test.cpp, line 136 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquNUwVCeHKBNo3dwsr:-LquNUwVCeHKBNo3dwss:bkc9rfl) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/test/persistence/dbschema_test.cpp#L136)):*
> ```cpp
> {
>     auto db = std::shared_ptr<RawDatabase>{new RawDatabase{"testIsNewDbTrue.db", {}, {}}};
>     bool bSuccess;
> ```

please initialize
___
*[test/persistence/dbschema_test.cpp, line 155 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquORA_1pRQlS2PWNRL:-LquORA_1pRQlS2PWNRM:b9ziewv) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/test/persistence/dbschema_test.cpp#L155)):*
> ```cpp
> }
> 
> void TestDbSchema::test1to2()
> ```

please add a comment, describing what is tested, especially the reason why we create friends with pending messages here. I know you're doing this to test the logic for the pending messages bug fix, but I'd like this documented here too, possibly with a link to the issue and the discussions.


<!-- Sent from Reviewable.io -->

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773)* status: :shipit: complete! 2 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 187 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquL5QaEZW_qGOEGvpU:-Lqu_itT1J2qhCx0loPC:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/src/persistence/history.cpp#L187)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

please initialize the variable
</blockquote></details>

Done.
___
*[src/persistence/history.cpp, line 670 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquPv0-8bkJSxXL-PE4:-Lqu_kLuExoDLkYrj5go:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/src/persistence/history.cpp#L670)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

There are multiple copies of the conversion, please extract into a function or make a constructor for MessageState (if C++ supports that)
</blockquote></details>

Done.
___
*[test/persistence/dbschema_test.cpp, line 136 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquNUwVCeHKBNo3dwsr:-Lqu_ldV8pOzFmsTxLa-:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/test/persistence/dbschema_test.cpp#L136)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

please initialize
</blockquote></details>

Done.
___
*[test/persistence/dbschema_test.cpp, line 155 at r7](https://reviewable.io/reviews/qtox/qtox/5773#-LquORA_1pRQlS2PWNRL:-Lqu_m8l1Dz41w5AlwfG:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/b1dd70a6070a74a7e898e210268305c5c8375f77/test/persistence/dbschema_test.cpp#L155)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

please add a comment, describing what is tested, especially the reason why we create friends with pending messages here. I know you're doing this to test the logic for the pending messages bug fix, but I'd like this documented here too, possibly with a link to the issue and the discussions.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 13 of 13 files at r9.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773#-:-LqurFQa8wZL012WPq_B:b-jbsf8z)* status: :shipit: complete! 2 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5773)* status: :shipit: complete! 2 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 118 at r12](https://reviewable.io/reviews/qtox/qtox/5773#-LrYqeAw7ro4HwzTkGly:-LrYqeAxB23AK0l7r6iV:b-sjgyuh) ([raw file](https://github.com/qtox/qtox/blob/cf63ae1a1b0c4494deca197f474de5c82162c7fd/src/persistence/history.cpp#L118)):*
> ```cpp
>         "LEFT JOIN faux_offline_pending ON history.id = faux_offline_pending.id "
>         "WHERE faux_offline_pending.id IS NULL "
>         "GROUP BY chat_id;");
> ```

I've fixed the SQL in this upgrade logic to fix the issue @sudden6 ran into when firs testing this PR. Previously I sorted the temporary table, then "grouped by" without a defined selection method - which is apparently undefined behaviour. Now the "group by" explicitly selects the max history id, e.g. the latest message. @sudden6 ran it again on the same database and didn't send a spurious message this time. I also noticed on one of my test profiles in one of my chats there was a pending message in the chat log after scrolling up. When using this updated version instead, that message was correctly marked as broken.


<!-- Sent from Reviewable.io -->
